res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js

```diff
@@ -442,23 +471,88 @@ TraktorS4MK2.registerOutputPackets = function() {
   Output3.addOutput("[EffectRack1_EffectUnit1]", "group_[Channel4]_enable", 0x0B, "B");
   Output3.addOutput("[EffectRack1_EffectUnit2]", "group_[Channel4]_enable", 0x0C, "B");
 
-  Output3.addOutput("[EffectRack1_EffectUnit1_Effect1]", "enabled", 0x02, "B");
-  Output3.addOutput("[EffectRack1_EffectUnit1_Effect2]", "enabled", 0x03, "B");
-  Output3.addOutput("[EffectRack1_EffectUnit1_Effect3]", "enabled", 0x04, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit1]", "!FXFocusButton", 0x01, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit1_Effect1]", "!FXButton1", 0x02, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit1_Effect2]", "!FXButton2", 0x03, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit1_Effect3]", "!FXButton3", 0x04, "B");
+
+  Output3.addOutput("[EffectRack1_EffectUnit2]", "!FXFocusButton", 0x0D, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit2_Effect1]", "!FXButton1", 0x0E, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit2_Effect2]", "!FXButton2", 0x0F, "B");
+  Output3.addOutput("[EffectRack1_EffectUnit2_Effect3]", "!FXButton3", 0x10, "B");
+
+  var effectFocusedCallback = function (focusedEffect, effectUnit, control) {
+    var regexMatch = effectUnit.match(script.effectUnitRegEx);
+    var effectUnitNumber = regexMatch[1];
+
+    for (var i = 1; i <= 3; i++) {
+      var effectString = "[EffectRack1_EffectUnit" + effectUnitNumber + "_Effect" + i +"]";
+
+      // disconnect old output callbacks
+      var oldConnection = TraktorS4MK2.controller.effect_button_output_connections[effectUnit][i];
+      if (oldConnection !== undefined) {
+        oldConnection.disconnect();
+      }
+
+      // Take care of soft takeover for the knobs being switched away from.
+      // previously_focused_effect is null on startup to avoid calling
+      // engine.softTakeoverIgnoreNextValue
+      var previouslyFocusedEffect =
+        TraktorS4MK2.controller.previously_focused_effect[effectUnit];
+
+      if (previouslyFocusedEffect === 0) {
+        engine.softTakeoverIgnoreNextValue(effectString, "meta");
+      } else if (previouslyFocusedEffect !== null) {
+        engine.softTakeoverIgnoreNextValue(
+          "[EffectRack1_EffectUnit" + effectUnitNumber + "_Effect" +
+          previouslyFocusedEffect + "]",
+          "parameter" + i);
+      }
+
+      var outputCallback = function (value, group, control) {
+        var led_value = 0x19;
+        if (value) {
+          led_value = 0x7F;
+        }
+        TraktorS4MK2.controller.setOutput(effectString, "!FXButton" + i, led_value,
+                                          !TraktorS4MK2.controller.freeze_lights);
```

Comment:
Try adding this line here:
`print("Sending output for (" + group + ", " + control + ") value " + value);`
Do you see this debugging output when pressing effects buttons? (You need to run Mixxx with the `--controllerDebug` command line argument to see it.)

Reply:
If you see debugging output from this then it may be that the focus/unfocusing logic is working but this part for sending the LED output is not correct.